File: src/main/java/ru/testproject/catalog/service/ProductService.java
package ru.testproject.catalog.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import ru.testproject.catalog.exception.NoDataFoundException;
import ru.testproject.catalog.model.Category;
import ru.testproject.catalog.model.Product;
import ru.testproject.catalog.repository.ProductRepository;

import java.util.List;

@Transactional
@Service
public class ProductService {
    private ProductRepository repository;
    private CategoryService categoryService;

    public ProductService(ProductRepository repository, CategoryService categoryService) {
        this.repository = repository;
        this.categoryService = categoryService;
    }

    @Transactional(readOnly = true, propagation = Propagation.SUPPORTS)
    public List<Product> getByCategory(String categoryName) {
        Category category = categoryService.getCategoryByName(categoryName);
        return category.getProducts();
    }

    public Product add(String productName, String categoryName) {
        Category category = categoryService.getCategoryByName(categoryName);
        Product product = new Product(productName);
        product.setCategory(category);
        category.getProducts().add(product);

        return repository.save(product);
    }

    public void move(String productName, String newCategoryName) {
        Category newCategory = categoryService.getCategoryByName(newCategoryName);
        Product product = repository.getByName(productName)
                .orElseThrow(() -> new NoDataFoundException(productName));
        Category oldCategory = categoryService.getCategoryByName(product.getCategory().getName());

        product.setCategory(newCategory);
        newCategory.getProducts().add(product);
        oldCategory.getProducts().remove(product);
    }

    public void delete(String productName) {
        Product product = repository.getByName(productName)
                .orElseThrow(() -> new NoDataFoundException(productName));
        product.getCategory().getProducts().remove(product);
        repository.delete(product);
    }
}

File: src/main/kotlin/ru/testproject/catalog/service/ProductService.kt
package ru.testproject.catalog.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import ru.testproject.catalog.repository.ProductRepository
import ru.testproject.catalog.service.CategoryService
import ru.testproject.catalog.model.Product
import ru.testproject.catalog.exception.NoDataFoundException

@Transactional
@Service
class ProductService(private val repository: ProductRepository, private val categoryService: CategoryService) {
    @Transactional(readOnly = true, propagation = Propagation.SUPPORTS)
    fun getByCategory(categoryName: String): List<Product> {
        val category = categoryService.getCategoryByName(categoryName)
        return category.products
    }

    fun add(productName: String, categoryName: String): Product {
        val category = categoryService.getCategoryByName(categoryName)
        val product = Product(productName)
        product.category = category
        category.products.add(product)
        return repository.save(product)
    }

    fun move(productName: String, newCategoryName: String) {
        val newCategory = categoryService.getCategoryByName(newCategoryName)
        val product = repository.getByName(productName)
            .orElseThrow { NoDataFoundException(productName) }

        val oldCategory = categoryService.getCategoryByName(product.category!!.name)
        product.category = newCategory
        newCategory.products.add(product)
        oldCategory.products.remove(product)
    }

    fun delete(productName: String) {
        val product = repository.getByName(productName)
            .orElseThrow { NoDataFoundException(productName) }

        product.category!!.products.remove(product)
        repository.delete(product)
    }
}
